lib/service_provider.rb

```diff
@@ -1,105 +1,63 @@
 require 'fingerprinter'
 
 class ServiceProvider
-  # currently acceptable encryption values are 'none' and 'aes256-cbc'
-  DEFAULT_ENCRYPTION = 'aes256-cbc'.freeze
-
   attr_reader :issuer
 
   def initialize(issuer)
     @issuer = issuer
   end
 
   def metadata
-    ignored_methods = [:metadata, :encrypt_responses?, :encryption_opts, :issuer]
-    metadata_methods = self.class.instance_methods(false) - ignored_methods
-
-    @metadata ||= metadata_methods.inject({}) do |hash, method|
-      hash.merge(:"#{method}" => send(method))
-    end
-  end
-
-  def acs_url
-    sp_attributes['acs_url']
-  end
-
-  def assertion_consumer_logout_service_url
-    sp_attributes['assertion_consumer_logout_service_url']
-  end
-
-  def sp_initiated_login_url
-    sp_attributes['sp_initiated_login_url']
-  end
-
-  def metadata_url
-    sp_attributes['metadata_url']
+    sp_attributes.merge(shared_attributes)
```

Comment:
Good catch! `merge! ` is 24x faster than `merge`: https://github.com/JuanitoFatas/fast-ruby#hashmerge-vs-hashmerge-code